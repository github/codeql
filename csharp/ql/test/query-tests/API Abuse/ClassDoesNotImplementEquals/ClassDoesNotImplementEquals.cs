using System;

// BAD
class Incorrect
{
    public static bool operator ==(Incorrect a, Incorrect b) => default(bool);
    public static bool operator !=(Incorrect a, Incorrect b) => !(a == b);
}

// GOOD
class Normal
{
}

// GOOD: provides Equals method.
class Correct
{
    public override bool Equals(object o) => default(bool);
    public static bool operator ==(Correct a, Correct b) => default(bool);
    public static bool operator !=(Correct a, Correct b) => !(a == b);
}

// BAD: needs to redefine Equals
class IncorrectOverrides : Correct
{
    public static bool operator ==(IncorrectOverrides a, IncorrectOverrides b) => default(bool);
    public static bool operator !=(IncorrectOverrides a, IncorrectOverrides b) => !(a == b);
}

// GOOD: abstract
abstract class Abstract
{
    public static bool operator ==(Abstract a, Abstract b) => default(bool);
    public static bool operator !=(Abstract a, Abstract b) => !(a == b);
}

class Program
{
    static void Main(string[] args)
    {
        // BAD: Call to missing equals method
        var b = new IncorrectOverrides().Equals(null);

        // GOOD:
        b = default(Abstract).Equals(null);
    }
}

// BAD: should also implement Equals.
class MyEquatable : IEquatable<MyEquatable>
{
    public bool Equals(MyEquatable other)
    {
        return value == other.value;
    }

    int value;
}

// GOOD: Implements Equals(object)
abstract class EqA<T> : IEquatable<T>
{
    public abstract bool Equals(T other);
    public override bool Equals(object other) { return other != null && GetType() == other.GetType() && Equals((T)other); }
}

// GOOD: handled by based class
class EqB : EqA<EqB>
{
    public override bool Equals(EqB other)
    {
        return true;
    }
}
