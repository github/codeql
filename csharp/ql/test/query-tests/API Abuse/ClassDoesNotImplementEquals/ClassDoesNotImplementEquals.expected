| ClassDoesNotImplementEquals.cs:4:7:4:15 | Incorrect | Class 'Incorrect' does not implement Equals(object), but it implements $@. | ClassDoesNotImplementEquals.cs:6:33:6:34 | == | operator == |
| ClassDoesNotImplementEquals.cs:24:7:24:24 | IncorrectOverrides | Class 'IncorrectOverrides' does not implement Equals(object), but $@. | ClassDoesNotImplementEquals.cs:42:17:42:53 | call to method Equals | 'Equals' is called on an instance of this class |
| ClassDoesNotImplementEquals.cs:24:7:24:24 | IncorrectOverrides | Class 'IncorrectOverrides' does not implement Equals(object), but it implements $@. | ClassDoesNotImplementEquals.cs:26:33:26:34 | == | operator == |
| ClassDoesNotImplementEquals.cs:50:7:50:17 | MyEquatable | Class 'MyEquatable' does not implement Equals(object), but it implements $@. | ClassDoesNotImplementEquals.cs:52:17:52:22 | Equals | IEquatable<MyEquatable>.Equals |
| ClassDoesNotImplementEqualsBad.cs:24:11:24:21 | GasolineCar | Class 'GasolineCar' does not implement Equals(object), but $@. | ClassDoesNotImplementEqualsBad.cs:38:38:38:54 | call to method Equals | 'Equals' is called on an instance of this class |
